### PR TITLE
Refactor Apriel2 cache, add Qwen2 converter, and conversation format for SFT

### PR DESCRIPTION
**Pull request by tscholak**
# Summary

## Apriel2 Cache Refactor

- Add methods to _AttentionCache and _SSMCache (reset, reorder, crop, batch_repeat, batch_select, is_initialized, batch_size)
- Add _iter_caches() helper to flatten stochastic layer dicts
- Simplify Apriel2Cache methods using new abstractions
- Fix sliding window attention mask sizes (cumulative_length tracking)
- Localize KDA tuple handling in _SSMCache

## Qwen2 Converter

- Add Qwen2/Qwen2.5 to Apriel2 config conversion (conversion/qwen2/ submodule)
- Add weight mapping plan for Qwen2 models
- Support Qwen-style bias pattern (QKV bias, no O bias)

## Per-Layer Bias Support

- Support weight-specific bias settings (query_layer.bias.enabled, etc.)
- Bias inheritance for stochastic mixer submixers
- Fix non-gated MLP handling (gate_proj only when gated=True)

## Surgery and Conversion Improvements

- Document monoidal structure with State/Partial Surgery/Transition Spec types
- Clarify algebraic laws (monoid operation, action law) in compose_configs
- Fix vision_encoder=None handling in converters
- Change to relative imports in apriel2 modules for portability

## SSM Layer Consolidation

- Remove DiscreteMamba2 (unused)
- Consolidate Mamba config (previously split across MambaConfig, MambaBaseConfig, Mamba2Config)
- Add varlen support for Mamba (via optional position_indices parameter)

## Conversation Format for SFT Data Preparation

- Add ConversationSourceConfig (type: conversation) for chat datasets (Tulu 3, ShareGPT, etc.)
- Fix O(n²) → O(n) tokenization via HuggingFace's return_assistant_tokens_mask
- Add tokenize_chat() method returning (tokens, train_mask) directly
- Add _mask_to_spans() helper to convert boolean mask to loss masking spans
- Fix chat template documentation: entire assistant turn must be inside {% generation %} markers (following SmolLM3 pattern)
- Add prepare_tulu3.yaml and train_supernet_qwen2.yaml training examples
- Document performance tuning (~8k tokens/s, ~61GB memory, ~25h for 1B tokens)

## Test plan

- Split cache tests into contract tests (vs HuggingFace) and Apriel2-specific tests
- Add integration tests for conversion: Qwen2 → Apriel2 → Supernet → Roundtrip
- Add compose_configs and plan_surgery tests
- Add expression plan tests (test_expr_plan.py)
- Add plan execution tests (test_plan_execution.py)
- Add tokenizer chat template validation and span extraction tests
- Add parameterized test_tokenize_chat with exact expected tokens and trainable indices
- Add test_mask_to_spans for span conversion
- Manual test: prepared Tulu 3 dataset (939K samples, ~6 minutes) with Qwen2 tokenizer
- Manual test: training run with ~8k tokens/s throughput

🤖 Generated with https://claude.com/claude-code

### CODE REVIEW

File: tests/data/test_tokenizer.py [hunk 1]
```diff
@@ -40,3 +40,120 @@ def test_tokenize_with_spans(common_tokenizer, spans, expected_token_spans, expe
...
+)
+def test_tokenize_chat(common_tokenizer, messages, expected_tokens, expected_trainable_indices):
+    common_tokenizer.tokenizer.chat_template = CHAT_TEMPLATE
+    tokens, train_mask = common_tokenizer.tokenize_chat(messages)
+    Assert.eq(tokens.tolist(), expected_tokens)
+    Assert.eq([i for i, m in enumerate(train_mask) if m], expected_trainable_indices)
+
+
+@pytest.mark.parametrize(
+    ("train_mask", "expected_loss_spans"),
```
Comment (RaymondLi0) [inline tests/data/test_tokenizer.py]: Fast-llm expects loss-masking spans in the format (begin, last), not (begin, last).
So we would need to add a -1 to all span-last values.
see: https://servicenow.github.io/Fast-LLM/recipes/instruction-finetuning/#step-2-format-the-dataset-into-a-chat-template and https://github.com/ServiceNow/Fast-LLM/blob/main/fast_llm/data/preparator/gpt_memmap/prepare.py#L222

**Review by RaymondLi0 — CHANGES_REQUESTED** [review]
I'd suggest to split the PR in smaller pieces to make it easier to understand.
I looked at the data-preparation changes: some fix needed to the loss-masking spans. Otherwise looks good!
